File: github.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"
)

const (
	graphqlEndpoint = "https://api.github.com/graphql"

	// The GitHub search API never returns more than 1000 results per query
	searchResultCap = 1000
)

const repoSearchQuery = `
query ($search: String!, $cursor: String) {
	search(type: REPOSITORY, query: $search, first: 100, after: $cursor) {
		pageInfo {
			hasNextPage
			endCursor
		}
		nodes {
			... on Repository {
				nameWithOwner
				url
				stargazerCount
				isPrivate
			}
		}
	}
}`

const issueSearchQuery = `
query ($search: String!, $cursor: String) {
	search(type: ISSUE, first: 100, query: $search, after: $cursor) {
		pageInfo {
			hasNextPage
			endCursor
		}
		nodes {
			... on Issue {
				number
				title
				url
				createdAt
				repository {
					nameWithOwner
					url
					isPrivate
				}
			}
			... on PullRequest {
				number
				title
				url
				createdAt
				state
				repository {
					nameWithOwner
					url
					isPrivate
				}
			}
		}
	}
}`

const discussionSearchQuery = `
query ($search: String!, $cursor: String) {
	search(type: DISCUSSION, first: 100, query: $search, after: $cursor) {
		pageInfo {
			hasNextPage
			endCursor
		}
		nodes {
			... on Discussion {
				number
				title
				url
				createdAt
				repository {
					nameWithOwner
					url
					isPrivate
				}
			}
		}
	}
}`

type client struct {
	token      string
	httpClient *http.Client
}

func newClient(token string) *client {
	return &client{
		token:      token,
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *client) famousRepos() ([]repo, error) {
	search := fmt.Sprintf(
		"user:%s stars:>=%d fork:false is:public sort:stars-desc",
		login,
		minRepoStars,
	)
	nodes, err := searchAll[repoNode](c, repoSearchQuery, search)
	if err != nil {
		return nil, err
	}
	repos := make([]repo, 0, len(nodes))
	for _, n := range nodes {
		if n.IsPrivate {
			continue
		}
		repos = append(repos, repo{Name: n.NameWithOwner, URL: n.URL, Stars: n.StargazerCount})
	}
	return repos, nil
}

func (c *client) createdPullRequests() ([]item, error) {
	search := fmt.Sprintf("is:pr is:public author:%s -user:%s sort:created-desc", login, login)
	return c.searchItems(issueSearchQuery, search)
}

func (c *client) participatedIssues() ([]item, error) {
	search := fmt.Sprintf("is:issue is:public involves:%s sort:updated-desc", login)
	return c.searchItems(issueSearchQuery, search)
}

func (c *client) participatedDiscussions() ([]item, error) {
	search := fmt.Sprintf("involves:%s sort:updated-desc", login)
	return c.searchItems(discussionSearchQuery, search)
}

func (c *client) searchItems(query, search string) ([]item, error) {
	nodes, err := searchAll[itemNode](c, query, search)
	if err != nil {
		return nil, err
	}
	items := make([]item, 0, len(nodes))
	for _, n := range nodes {
		// Guards against private results when running with a broadly scoped local token
		if n.Repository.IsPrivate {
			continue
		}
		items = append(items, item{
			Number:    n.Number,
			Title:     n.Title,
			URL:       n.URL,
			State:     strings.ToLower(n.State),
			CreatedAt: n.CreatedAt,
			RepoName:  n.Repository.NameWithOwner,
			RepoURL:   n.Repository.URL,
		})
	}
	return items, nil
}

type pageInfo struct {
	HasNextPage bool   `json:"hasNextPage"`
	EndCursor   string `json:"endCursor"`
}

type repoNode struct {
	NameWithOwner  string `json:"nameWithOwner"`
	URL            string `json:"url"`
	StargazerCount int    `json:"stargazerCount"`
	IsPrivate      bool   `json:"isPrivate"`
}

type itemNode struct {
	Number     int       `json:"number"`
	Title      string    `json:"title"`
	URL        string    `json:"url"`
	CreatedAt  time.Time `json:"createdAt"`
	State      string    `json:"state"`
	Repository struct {
		NameWithOwner string `json:"nameWithOwner"`
		URL           string `json:"url"`
		IsPrivate     bool   `json:"isPrivate"`
	} `json:"repository"`
}

func searchAll[T any](c *client, query, search string) ([]T, error) {
	var all []T
	var cursor *string
	for {
		var result struct {
			Search struct {
				PageInfo pageInfo `json:"pageInfo"`
				Nodes    []T      `json:"nodes"`
			} `json:"search"`
		}
		variables := map[string]any{"search": search, "cursor": cursor}
		if err := c.query(query, variables, &result); err != nil {
			return nil, err
		}
		all = append(all, result.Search.Nodes...)
		if !result.Search.PageInfo.HasNextPage || len(all) >= searchResultCap {
			return all, nil
		}
		cursor = &result.Search.PageInfo.EndCursor
	}
}

func (c *client) query(query string, variables map[string]any, out any) error {
	payload, err := json.Marshal(map[string]any{"query": query, "variables": variables})
	if err != nil {
		return err
	}
	req, err := http.NewRequest(http.MethodPost, graphqlEndpoint, bytes.NewReader(payload))
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+c.token)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("graphql request returned status %s", resp.Status)
	}

	var envelope struct {
		Data   json.RawMessage `json:"data"`
		Errors []struct {
			Message string `json:"message"`
		} `json:"errors"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&envelope); err != nil {
		return err
	}
	if len(envelope.Errors) > 0 {
		return fmt.Errorf("graphql error: %s", envelope.Errors[0].Message)
	}
	return json.Unmarshal(envelope.Data, out)
}

File: lib/github.js
const graphqlEndpoint = "https://api.github.com/graphql";
const requestTimeout = 30_000;

// The GitHub search API never returns more than 1000 results per query
const searchResultCap = 1000;

const repoSearchQuery = `
query ($search: String!, $cursor: String) {
	search(type: REPOSITORY, query: $search, first: 100, after: $cursor) {
		pageInfo {
			hasNextPage
			endCursor
		}
		nodes {
			... on Repository {
				nameWithOwner
				url
				stargazerCount
				isPrivate
			}
		}
	}
}`;

const issueSearchQuery = `
query ($search: String!, $cursor: String) {
	search(type: ISSUE, first: 100, query: $search, after: $cursor) {
		pageInfo {
			hasNextPage
			endCursor
		}
		nodes {
			... on Issue {
				number
				title
				url
				createdAt
				repository {
					nameWithOwner
					url
					isPrivate
					owner {
						login
						avatarUrl(size: 64)
					}
				}
			}
			... on PullRequest {
				number
				title
				url
				createdAt
				state
				repository {
					nameWithOwner
					url
					isPrivate
					owner {
						login
						avatarUrl(size: 64)
					}
				}
			}
		}
	}
}`;

const discussionSearchQuery = `
query ($search: String!, $cursor: String) {
	search(type: DISCUSSION, first: 100, query: $search, after: $cursor) {
		pageInfo {
			hasNextPage
			endCursor
		}
		nodes {
			... on Discussion {
				number
				title
				url
				createdAt
				repository {
					nameWithOwner
					url
					isPrivate
					owner {
						login
						avatarUrl(size: 64)
					}
				}
			}
		}
	}
}`;

const contributionCalendarQuery = `
query ($login: String!) {
	user(login: $login) {
		contributionsCollection {
			contributionCalendar {
				totalContributions
				weeks {
					contributionDays {
						date
						contributionCount
						contributionLevel
					}
				}
			}
		}
	}
}`;

const contributionLevels = {
	FIRST_QUARTILE: 1,
	SECOND_QUARTILE: 2,
	THIRD_QUARTILE: 3,
	FOURTH_QUARTILE: 4,
};

export class GitHubClient {
	#token;
	#login;

	constructor(token, login) {
		this.#token = token;
		this.#login = login;
	}

	async famousRepos(minStars) {
		const search = `user:${this.#login} stars:>=${minStars} fork:false is:public sort:stars-desc`;
		const nodes = await this.#searchAll(repoSearchQuery, search);
		const prefix = `${this.#login}/`;
		return nodes
			.filter((n) => !n.isPrivate)
			.map((n) => ({
				name: n.nameWithOwner.startsWith(prefix)
					? n.nameWithOwner.slice(prefix.length)
					: n.nameWithOwner,
				url: n.url,
				stars: n.stargazerCount,
			}));
	}

	createdPullRequests() {
		const search = `is:pr is:public author:${this.#login} -user:${this.#login} sort:created-desc`;
		return this.#searchItems(issueSearchQuery, search);
	}

	participatedIssues() {
		const search = `is:issue is:public involves:${this.#login} sort:updated-desc`;
		return this.#searchItems(issueSearchQuery, search);
	}

	participatedDiscussions() {
		const search = `involves:${this.#login} sort:updated-desc`;
		return this.#searchItems(discussionSearchQuery, search);
	}

	async contributionCalendar() {
		const data = await this.#query(contributionCalendarQuery, { login: this.#login });
		const calendar = data.user.contributionsCollection.contributionCalendar;
		return {
			total: calendar.totalContributions,
			weeks: calendar.weeks.map((week) =>
				week.contributionDays.map((day) => ({
					date: parseDate(day.date),
					count: day.contributionCount,
					level: contributionLevels[day.contributionLevel] ?? 0,
				})),
			),
		};
	}

	async #searchItems(query, search) {
		const nodes = await this.#searchAll(query, search);
		return (
			nodes
				// Guards against private results when running with a broadly scoped local token
				.filter((n) => !n.repository.isPrivate)
				.map((n) => ({
					number: n.number,
					title: n.title,
					url: n.url,
					state: (n.state ?? "").toLowerCase(),
					createdAt: new Date(n.createdAt),
					repoName: n.repository.nameWithOwner,
					repoUrl: n.repository.url,
					repoOwnerLogin: n.repository.owner.login,
					repoOwnerAvatarUrl: n.repository.owner.avatarUrl,
				}))
		);
	}

	async #searchAll(query, search) {
		const all = [];
		let cursor = null;
		for (;;) {
			const { search: result } = await this.#query(query, { search, cursor });
			all.push(...result.nodes);
			if (!result.pageInfo.hasNextPage || all.length >= searchResultCap) {
				return all;
			}
			cursor = result.pageInfo.endCursor;
		}
	}

	async #query(query, variables) {
		const response = await fetch(graphqlEndpoint, {
			method: "POST",
			headers: {
				Authorization: `Bearer ${this.#token}`,
				"Content-Type": "application/json",
			},
			body: JSON.stringify({ query, variables }),
			signal: AbortSignal.timeout(requestTimeout),
		});
		if (!response.ok) {
			throw new Error(`graphql request returned status ${response.status} ${response.statusText}`);
		}
		const { data, errors } = await response.json();
		if (errors?.length) {
			throw new Error(`graphql error: ${errors[0].message}`);
		}
		return data;
	}
}

function parseDate(value) {
	const date = new Date(value);
	if (Number.isNaN(date.getTime())) {
		throw new Error(`parse contribution date "${value}"`);
	}
	return date;
}
